File: java/41.first-missing-positive.java
/*
 * @lc app=leetcode id=41 lang=java
 *
 * [41] First Missing Positive
 */
class Solution {
    public int firstMissingPositive(int[] nums) {
        int i;
        for (i = 0; i < nums.length; i++) {
            while (nums[i] > 0 && nums[i] <= nums.length && nums[i] != nums[nums[i] - 1]) {
                int tmp = nums[i];
                nums[i] = nums[tmp - 1];
                nums[tmp - 1] = tmp;
            }
        }
        for (i = 0; i < nums.length; i++) {
            if (nums[i] != i + 1)
                return i + 1;
        }
        return i + 1;
    }
}

File: cpp/41_First Missing Positive.cpp
// put postive n at the pos[n-1].
// if n = idx, n is not a missing positive
class Solution {
public:
    int firstMissingPositive(vector<int>& nums) {
        for(auto n : nums){
        	while(n > 0 && n <= nums.size() && n != nums[n-1]){
        		swap(n, nums[n-1]);
        	}
        }
        for(int i = 0; i < nums.size(); ++i){
            //cout<<nums[i];
        	if(nums[i] != i+1){
        		return i+1;
        	}
        }
        return nums.size() + 1;
    }
};